base/event.jl

```diff
@@ -1,5 +1,85 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
+## thread/task locking abstraction
+
+"""
+    AbstractLock
+
+Abstract supertype describing types that
+implement the synchronization primitives:
+[`lock`](@ref), [`trylock`](@ref), [`unlock`](@ref), and [`islocked`](@ref).
+"""
+abstract type AbstractLock end
+function lock end
+function unlock end
+function trylock end
+function islocked end
+unlockall(l::AbstractLock) = unlock(l) # internal function for implementing `wait`
+relockall(l::AbstractLock, token::Nothing) = lock(l) # internal function for implementing `wait`
+assert_havelock(l::AbstractLock) = assert_havelock(l, Threads.threadid())
+assert_havelock(l::AbstractLock, tid::Integer) =
+    (islocked(l) && tid == Threads.threadid()) ? nothing : error("concurrency violation detected")
+assert_havelock(l::AbstractLock, tid::Task) =
+    (islocked(l) && tid === current_task()) ? nothing : error("concurrency violation detected")
+assert_havelock(l::AbstractLock, tid::Nothing) = error("concurrency violation detected")
+
+"""
+    AlwaysLockedST
+
+This struct does not implement a real lock, but instead
+pretends to be always locked on the original thread it was allocated on,
+and simply ignores all other interactions.
+It also does not synchronize tasks; for that use a [`CooperativeLock`](@ref) or a
+real lock such as [`RecursiveLock`](@ref).
+This can be used in the place of a real lock to, instead, simply and cheaply assert
+that the operation is only occurring on a single thread.
+And is thus functionally equivalent to allocating a real, recursive, task-unaware lock
+immediately calling `lock` on it, and then never calling a matching `unlock`,
+except that calling `lock` from another thread will throw a concurrency violation exception.
+"""
+struct AlwaysLockedST <: AbstractLock
+    ownertid::Int16
+    AlwaysLockedST() = new(Threads.threadid())
+end
+assert_havelock(l::AlwaysLockedST) = assert_havelock(l, l.ownertid)
+lock(l::AlwaysLockedST) = assert_havelock(l)
+unlock(l::AlwaysLockedST) = assert_havelock(l)
+trylock(l::AlwaysLockedST) = l.ownertid == Threads.threadid()
+islocked(::AlwaysLockedST) = true
+
+"""
+    CooperativeLock
+
+An optimistic lock for cooperative tasks, which can be used cheaply to check for missing
+lock/unlock guards around `wait`, in the trivial (conflict-free, yield-free, single-threaded, non-recursive) case,
+without paying the cost for a full RecursiveLock.
```

Comment:
Should this lock be used for anything other than as a stepping stone on the upgrade path to full multi-threadedness? If not, can this docstring be more strongly worded to discourage misuse? :)

Reply:
No, it shouldn't. It was part of the scaffolding of this PR, and should have been removed now (it's unused). Good catch.